openhtf/output/callbacks/junit.py

```diff
@@ -0,0 +1,85 @@
+"""Module for outputting test record to JUNIT-formatted files."""
+
+import base64
+import six
+from junit_xml import TestSuite, TestCase
+
+from openhtf.output import callbacks
+from openhtf.util import data
+
+
+class OutputToJUNIT(callbacks.OutputToFile):
+  """Return an output callback that writes JUNIT Test Records.
+  Example filename_patterns might be:
+    '/data/test_records/{dut_id}.{metadata[test_name]}.xml', indent=4)) or
+    '/data/test_records/%(dut_id)s.%(start_time_millis)s'
+  To use this output mechanism:
+    test = openhtf.Test(PhaseOne, PhaseTwo)
+    test.add_output_callback(openhtf.output.callbacks.OutputToJUNIT(
+        '/data/test_records/{dut_id}.{metadata[test_name]}.xml'))
+
+  Args:
+    filename_pattern: A format string specifying the filename to write to,
+      will be formatted with the Test Record as a dictionary.  May also be a
+      file-like object to write to directly.
+    inline_attachments: Whether attachments should be included inline in the
+      output. Set to False if you expect to have large binary attachments. If
+      True (the default), then attachments are base64 encoded to allow for
+      binary data that's not supported by JUNIT directly.
+  """
+
+  def __init__(self, filename_pattern=None, **kwargs):
+    super(OutputToJUNIT, self).__init__(filename_pattern)
+    self.test_cases = []
+
+  def serialize_test_record(self, test_record):
+    dict_test_record = self.convert_to_dict(test_record)
+    test_case = None
+
+    for phase in dict_test_record["phases"]:
+      output = []
+      for _, phase_data in phase["measurements"].items():
+        log_output = []
+
+        for log_data in dict_test_record["log_records"]:
+          if phase["name"] in log_data["logger_name"]:
+            log_output.append(log_data["message"])
+
+        output.extend(["name: " + phase_data["name"],
+                       "validators: " + str(phase_data["validators"]),
+                       "measured_value: " + str(phase_data["measured_value"]),
+                       "outcome: " + phase_data["outcome"],
+                       "log: " + "\n".join(log_output), "\n"])
```

Comment:
I was having an issue with this throwing an exception because there was no "validators" entry in the phase_data dictionary for a measurement generated by a monitor. The change below resolved that issue with me.

```
        if "name" in phase_data:
          output.extend(["name: " + phase_data["name"]])
        if "validators" in phase_data:
          output.extend(["validators: " + str(phase_data["validators"])])
        if "measured_value" in phase_data:
          output.extend(["measured_value: " + str(phase_data["measured_value"])])
        if "outcome" in phase_data:
          output.extend(["outcome: " + phase_data["outcome"]])
        output.extend(["log: " + "\n".join(log_output), "\n"])
```